pkg/client/chunk/expected.go

```diff
@@ -0,0 +1,152 @@
+			return nil, status.Errorf(
+				codes.ResourceExhausted,
+				"unable to chunk proto.CheckinExpected the unit %s is larger than max (%d vs. %d)",
+				u.Id, bySize[i].size, maxSize)
+		}
+	}
+
+	// sort the smallest units first, this ensures that the first chunk that includes extra
+	// fields uses the smallest unit to ensure that it all fits
+	slices.SortStableFunc(bySize, func(a, b expectedBySize) int {
+		return a.size - b.size
+	})
+
+	// first message all fields are set; except units is made smaller
+	m := shallowCopyCheckinExpected(msg)
+	m.Units = make([]*proto.UnitExpected, 0, 1)
+	m.Units = append(m.Units, bySize[0].unit)
+	m.UnitsTimestamp = timestamppb.New(options.timestamp)
+	s = protobuf.Size(m)
+	if s >= maxSize {
+		// not possible even for the first chunk to fit
+		return nil, status.Errorf(
+			codes.ResourceExhausted,
+			"unable to chunk proto.CheckinExpected the first chunk with unit %s is larger than max (%d vs. %d)",
+			m.Units[0].Id, s, maxSize)
+	}
+
+	// keep adding units until it doesn't fit
+	for nextUnit := 1; s < maxSize && nextUnit < len(bySize); nextUnit++ {
+		us := bySize[nextUnit]
+		if s+us.size < maxSize {
+			// unit fits add it
+			m.Units = append(m.Units, us.unit)
+			s += us.size
+		} else {
+			// doesn't fit, create a new chunk
+			msgs = append(msgs, m)
+			m = &proto.CheckinExpected{}
+			m.UnitsTimestamp = timestamppb.New(options.timestamp)
+			m.Units = make([]*proto.UnitExpected, 0, 1)
+			m.Units = append(m.Units, us.unit)
+			s = protobuf.Size(m)
+		}
+	}
+	msgs = append(msgs, m)
+
+	// all chunks created, create the empty chunk
+	m = &proto.CheckinExpected{}
+	m.UnitsTimestamp = timestamppb.New(options.timestamp)
+	m.Units = make([]*proto.UnitExpected, 0)
+	msgs = append(msgs, m)
+	return msgs, nil
+}
+
+// CheckinExpectedReceiver provides a Recv interface to receive proto.CheckinExpected messages.
+type CheckinExpectedReceiver interface {
+	Recv() (*proto.CheckinExpected, error)
+}
+
+// RecvExpected handles the receiving of chunked proto.CheckinObjected.
+func RecvExpected(recv CheckinExpectedReceiver) (*proto.CheckinExpected, error) {
+	var first *proto.CheckinExpected
+	for {
+		msg, err := recv.Recv()
+		if err != nil {
+			return nil, err
+		}
+		if msg.UnitsTimestamp == nil {
+			// all included in a single message
+			return msg, nil
+		}
+		if first == nil {
+			// first message in batch
+			first = msg
+		} else if first.UnitsTimestamp.AsTime() != msg.UnitsTimestamp.AsTime() {
+			// only used if the new timestamp is newer
+			if first.UnitsTimestamp.AsTime().After(msg.UnitsTimestamp.AsTime()) {
+				// not newer so we ignore the message
+				continue
+			}
+			// different batch; restart
+			first = msg
+		}
+		if len(msg.Units) == 0 {
+			// ending match message
+			return first, nil
+		}
+		if first != msg {
+			first.Units = append(first.Units, msg.Units...)
+		}
+	}
+}
+
+func shallowCopyCheckinExpected(msg *proto.CheckinExpected) *proto.CheckinExpected {
```

Comment:
`proto.Clone` is a deep copy but has the advantage that we don't have to keep this function up to date if we add a top level message field.

Is there a test we can write that would fail if you forgot to update this function (and the observed version of it)?